testing/unittests/jthreadtests.cpp

```diff
@@ -0,0 +1,591 @@
+
+        // This should block for approximately 100ms (start delay) then throw exception
+        try
+        {
+            pool->start(&lifespan2seconds, "Thread4", 100); // timeout after 100ms
+            CPPUNIT_FAIL("Start should have thrown an exception as no slots would be available when start delay specified");
+        }
+        catch (IException *e)
+        {
+            auto throttleEndTime = std::chrono::high_resolution_clock::now();
+            auto throttleDuration = std::chrono::duration_cast<std::chrono::milliseconds>(throttleEndTime - throttleStartTime);
+
+            CPPUNIT_ASSERT(throttleDuration.count() >= 80 && throttleDuration.count() <= 150);
+
+            StringBuffer msg;
+            e->errorMessage(msg);
+            CPPUNIT_ASSERT_EQUAL_STR("No threads available in pool ThrottledTestPool", msg.str());
+            e->Release();
+        }
+
+        auto longThrottleStartTime = std::chrono::high_resolution_clock::now();
+
+        // This should block for the full 500ms delay then not throw exception
+        try
+        {
+            pool->start(&lifespan2seconds, "Thread4"); // Will use default delay of 500ms
+        }
+        catch (IException *e)
+        {
+            StringBuffer msg("Exception encountered: ");
+            e->errorMessage(msg);
+            CPPUNIT_FAIL(msg.str());
+            e->Release();
+        }
+        auto longThrottleEndTime = std::chrono::high_resolution_clock::now();
+        auto longThrottleDuration = std::chrono::duration_cast<std::chrono::milliseconds>(longThrottleEndTime - longThrottleStartTime);
+
+        CPPUNIT_ASSERT(longThrottleDuration.count() >= 450 && longThrottleDuration.count() <= 600);
+
+        // Wait for threads to complete
+        std::this_thread::sleep_for(std::chrono::seconds(3));
+        CPPUNIT_ASSERT_EQUAL(0U, pool->runningCount());
+        CPPUNIT_ASSERT_EQUAL(4U, threadStartedCount.load());
+        CPPUNIT_ASSERT_EQUAL(4U, threadCompletedCount.load());
+
+        // Test: No throttle delay is introduced after pool falls below defaultmax
+        auto afterDropStartTime = std::chrono::high_resolution_clock::now();
+
+        pool->start(&lifespan2seconds, "Thread5");
+        pool->start(&lifespan2seconds, "Thread6");
+
+        auto afterDropEndTime = std::chrono::high_resolution_clock::now();
+        auto afterDropDuration = std::chrono::duration_cast<std::chrono::milliseconds>(afterDropEndTime - afterDropStartTime);
+
+        // Should start immediately without throttling
+        CPPUNIT_ASSERT(afterDropDuration.count() < 100);
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        CPPUNIT_ASSERT_EQUAL(2U, pool->runningCount());
+
+        // Test: Pool was above, fell below, and is now at defaultmax again
+        pool->start(&lifespan2seconds, "Thread7"); // Now at capacity (3 threads)
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
```

Comment:
not really sure what this is for - to give 1ms time for them to begin? too risky to guarantee this way.
See other comments, need to use semaphores really to know.